File: 0290-word-pattern/0290-word-pattern.py
class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()
        if len(pattern) != len(words):
            return False
            
        charLastSeen = [0] * 26
        wordLastSeen = {}
        
        for i in range(len(pattern)):
            charIdx = ord(pattern[i]) - 97
            word = words[i]
            
            lastCharPos = charLastSeen[charIdx]
            lastWordPos = wordLastSeen.get(word, 0)
            
            if lastCharPos != lastWordPos:
                return False
                
            charLastSeen[charIdx] = i + 1
            wordLastSeen[word] = i + 1
            
        return True

File: 0290-word-pattern/0290-word-pattern.js
function wordPattern(pattern, s) {
    const words = s.split(" ");
    if (pattern.length !== words.length) return false;

    const charLastSeen = new Int32Array(26);
    const wordLastSeen = new Map();

    for (let i = 0; i < pattern.length; i++) {
        const charIdx = pattern.charCodeAt(i) - 97;
        const word = words[i];

        const lastCharPos = charLastSeen[charIdx];
        const lastWordPos = wordLastSeen.get(word) || 0;

        if (lastCharPos !== lastWordPos) return false;

        charLastSeen[charIdx] = i + 1;
        wordLastSeen.set(word, i + 1);
    }
    return true;
}
